### PR TITLE
Refactor data types and api response types

### PR DESCRIPTION
**Pull request by riyajaggi**
### Summary 

This pull request is the first step to refactor IDOL endpoints for member and teams APIs by defining all data types in `DataTypes.d.ts` and all API response types in `APITypes.d.ts`. The endpoints return the status code and either the error message or the data on success. The error messages and response codes are also updated as per [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). 

**Error codes for members:**
200 - successful
401 - session user doesn't exist in db
404 - member doesn't exist in db
403 - user does not have permission to edit
400 - undefined email
500 - internal errors due to the db (various reasons)

**Error codes for teams:**
200 - successful
404 - members don't exist in db _or_ no team exists with the entered uuid
403 - user does not have permission to edit
400 - malformed teams or members on POST
500 - internal errors due to db (various reasons)

### Test Plan 

Need to test locally (this includes tests from #9):
- check if the user can update their general info correctly
- check if user can only update their information and not another user's (unless their role is admin or lead)
- check if only admins/leads can update the role
- check if only admins/leads can update the first name/last name
- check if only the user themselves or the admins/leads get the member info
- check if all old member/team endpoints work as intended with the new changes to response types

### Notes 

- For the member type, should the fields `role`, `subteam` and `other_subteams` be updated to types `role`, `Team` and `Team[]` respectively, instead of `string`?
- Not sure if this is needed, but we aren't checking if the request body matches the` Member` type in member endpoints.

### Breaking Changes  

This shouldn't affect any frontend implementation, since it just involves refactoring backend code.

### CODE REVIEW

**Review by henry-li-06 — COMMENTED** [review]
Good work! It looks like most of our refactoring for the data and api types looks mostly similar. Using intersections for the api types seems like a really good idea!

Otherwise I think that like you were saying, we can make the /deleteMember and /deleteTeam endpoints into DELETE requests to better follow REST practices. Finally, I think for some of the error statuses you have 404 for when email of the user doesn't exist in our database. I think something like a 401 might be better suited since the current user isn't necessarily the requested resource that's missing. What do you think about this?

File: src/api.ts [hunk 1]
```diff
@@ -92,23 +113,41 @@ router.post('/logout', (req, res) => {
 router.get('/allRoles', getAllRoles);
 
 // Members
-router.get('/allMembers', allMembers);
-router.post('/setMember', setMember);
-router.post('/deleteMember', deleteMember);
+router.get('/allMembers', async (req, res) => {
+  let handled = await allMembers(req, res); res.status(handled.status).json(handled);
```
Comment (henry-li-06) [inline src/api.ts:117]: I think this line would be better off as two separate lines for each of the routes (lines 117, 120, 123, etc.)
i.e.<br /> ` let handled = ...`<br /> `res.status(...`

File: src/memberAPI.ts [hunk 2]
```diff
@@ -1,70 +1,163 @@
...
+    } else {
+      let canEdit: boolean = PermissionsManager.canEditMembers(member.role);
+      let memberEmail: string = req.params.email;
+      if (!canEdit && memberEmail !== req.session.email) {
+        return {
+          status: 403,
+          error: 'User with email: ' + req.session.email + ' does not have permission to get members!'
+        };
       }
+      let getmember: Member = (await db.doc('members/' + memberEmail).get()).data() as Member;
```
Comment (henry-li-06) [inline src/memberAPI.ts:134]: This variable name could probably just be `member`. `getmember` seems a bit confusing to me.

File: src/memberAPI.ts [hunk 3]
```diff
@@ -1,70 +1,163 @@
...
-          res.status(200).json({ error: "Couldn't delete user for reason: " + reason });
-        });
+        if ((req.body.role || req.body.first_name || req.body.last_name) &&
+          !canEdit) {
+          return {
+            status: 403,
+            error: 'User with email: ' + req.session.email + ' does not have permission to edit member name or roles!',
+          };
+        }
+        db.doc('members/' + req.body.email)
```
Comment (henry-li-06) [inline src/memberAPI.ts:95]: When you call the database to update the member information, there needs to be a return outside of this expression. <br />
i.e. something like 
```
let response = db.doc(...
return response;
```
Otherwise the function isn't returning anything since the returns in the `.then()` are just returning a value in the expression.

**Comment by riyajaggi** [comment]
> Good work! It looks like most of our refactoring for the data and api types looks mostly similar. Using intersections for the api types seems like a really good idea!
> 
> Otherwise I think that like you were saying, we can make the /deleteMember and /deleteTeam endpoints into DELETE requests to better follow REST practices. Finally, I think for some of the error statuses you have 404 for when email of the user doesn't exist in our database. I think something like a 401 might be better suited since the current user isn't necessarily the requested resource that's missing. What do you think about this?

**Comment by riyajaggi** [comment]
Thank you @henry-li-06 for your comments! I update the /deleteMember and /deleteTeam endpoints into DELETE. I agree with your remark about error codes, 401 would be better suited since that session user is not authorized to see our documents since they are not in idol db.

File: src/memberAPI.ts (hunk 3 above)
Comment (riyajaggi) [reply src/memberAPI.ts]: Thank you for pointing this out! I have updated the functions to better reflect the responses returned.

**Review by JBoss925 — APPROVED** [review]
LGTM!

File: .eslintrc.json [hunk 4]
```diff
@@ -21,15 +21,6 @@
...
-            "error",
-            "ignorePackages",
-            {
-                "js": "never",
-                "jsx": "never",
-                "ts": "never",
-                "tsx": "never"
-            }
-        ]
+        "import/extensions": "off"
```
Comment (SamChou19815) [inline .eslintrc.json:24]: why this change?

File: src/memberAPI.ts [hunk 5]
```diff
@@ -2,86 +2,103 @@ import { Request, Response } from 'express';
...
       .get()
-      .then((vals) => vals.docs.map((doc) => doc.data()) as Member[]);
-    return { members, status: 200 };
+      .then((vals) => {
+        return vals.docs.map((doc) => doc.data()) as Member[];
+      });
+    return {
+      status: 200,
+      members
+    };
```
Comment (SamChou19815) [inline src/memberAPI.ts:22]: please avoid formatting only changes like this

File: src/memberAPI.ts [hunk 6]
```diff
@@ -2,86 +2,103 @@ import { Request, Response } from 'express';
...
+          status: 200,
+          member: req.body
+        };
+      })
+      .catch((reason) => {
+        return {
+          status: 500,
+          error: `Couldn't edit member ${req.body.email} for reason: ${reason}`
+        };
+      });
```
Comment (SamChou19815) [inline src/memberAPI.ts:69]: another formatting only change

**Review by SamChou19815 — COMMENTED** [review]
There are a lot of formatting only changes like adding new indentation and switch order of fields. They are unnecessary and increase the chances of merge conflicts with other's work.
